File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/domain/sync/LocalFolderSyncStorage.kt
/*
 * BMLibrarian Lite - Biomedical Literature Research Tool
 * Copyright (C) 2024-2025 Dr Horst Herb
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.bmlibrarian.factchecker.domain.sync

import android.net.Uri
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Local folder-based sync storage implementation.
 *
 * Works with any user-selected folder, enabling sync via:
 * - Dropbox, Google Drive, iCloud, OneDrive (folder sync)
 * - Syncthing (P2P sync)
 * - Network shares (SMB/NFS)
 * - Local device storage (no sync)
 *
 * All file operations are performed on the IO dispatcher for proper threading.
 * Writes are atomic (write to temp file, then rename) to prevent corruption.
 *
 * @property rootPath Absolute path to the sync root folder
 *
 * @see <a href="doc/cross_platform/sync_protocol.md">Sync Protocol Specification</a>
 */
class LocalFolderSyncStorage(
    private val rootPath: String
) : SyncStorage {

    companion object {
        private const val TAG = "LocalFolderSyncStorage"
        private const val TEMP_FILE_PREFIX = ".tmp_"
    }

    private val rootFile = File(rootPath)

    init {
        require(rootPath.isNotBlank()) { "Root path cannot be blank" }
    }

    /**
     * Resolves a relative path to an absolute File.
     */
    private fun resolvePath(relativePath: String): File {
        val normalizedPath = relativePath.trimStart('/')
        return if (normalizedPath.isEmpty()) {
            rootFile
        } else {
            File(rootFile, normalizedPath)
        }
    }

    override suspend fun listFiles(path: String): List<SyncFileInfo> = withContext(Dispatchers.IO) {
        val dir = resolvePath(path)

        if (!dir.exists()) {
            throw SyncStorageException.DirectoryNotFound(path)
        }

        if (!dir.isDirectory) {
            throw SyncStorageException.DirectoryNotFound("$path is not a directory")
        }

        try {
            dir.listFiles()
                ?.filter { it.isFile && !it.name.startsWith(".") }
                ?.map { file ->
                    SyncFileInfo(
                        name = file.name,
                        path = file.relativeTo(rootFile).path,
                        size = file.length(),
                        modifiedAt = file.lastModified()
                    )
                }
                ?.sortedByDescending { it.modifiedAt }
                ?: emptyList()
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        }
    }

    override suspend fun listDirectories(path: String): List<String> = withContext(Dispatchers.IO) {
        val dir = resolvePath(path)

        if (!dir.exists()) {
            throw SyncStorageException.DirectoryNotFound(path)
        }

        if (!dir.isDirectory) {
            throw SyncStorageException.DirectoryNotFound("$path is not a directory")
        }

        try {
            dir.listFiles()
                ?.filter { it.isDirectory && !it.name.startsWith(".") }
                ?.map { it.name }
                ?.sorted()
                ?: emptyList()
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        }
    }

    override suspend fun readFile(path: String): ByteArray = withContext(Dispatchers.IO) {
        val file = resolvePath(path)

        if (!file.exists()) {
            throw SyncStorageException.FileNotFound(path)
        }

        if (!file.isFile) {
            throw SyncStorageException.FileNotFound("$path is not a file")
        }

        try {
            file.readBytes()
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        } catch (e: IOException) {
            throw SyncStorageException.IOError("Failed to read $path", e)
        }
    }

    override suspend fun writeFile(data: ByteArray, path: String) = withContext(Dispatchers.IO) {
        val file = resolvePath(path)

        try {
            // Ensure parent directories exist
            file.parentFile?.let { parent ->
                if (!parent.exists()) {
                    if (!parent.mkdirs()) {
                        throw SyncStorageException.IOError("Failed to create directory: ${parent.path}")
                    }
                }
            }

            // Atomic write: write to temp file, then rename
            val tempFile = File(file.parentFile, "$TEMP_FILE_PREFIX${file.name}")
            try {
                tempFile.writeBytes(data)

                // Rename temp file to final destination
                if (!tempFile.renameTo(file)) {
                    // Fallback: copy and delete if rename fails (cross-filesystem)
                    tempFile.copyTo(file, overwrite = true)
                    tempFile.delete()
                }

                Log.d(TAG, "Wrote ${data.size} bytes to $path")
            } catch (e: Exception) {
                // Clean up temp file on failure
                tempFile.delete()
                throw e
            }
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        } catch (e: SyncStorageException) {
            throw e
        } catch (e: IOException) {
            throw SyncStorageException.IOError("Failed to write $path", e)
        }
    }

    override suspend fun deleteFile(path: String) = withContext(Dispatchers.IO) {
        val file = resolvePath(path)

        if (!file.exists()) {
            // Idempotent: no error if file doesn't exist
            Log.d(TAG, "Delete ignored (file not found): $path")
            return@withContext
        }

        try {
            if (file.isFile) {
                if (!file.delete()) {
                    throw SyncStorageException.IOError("Failed to delete $path")
                }
                Log.d(TAG, "Deleted file: $path")
            }
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        }
    }

    override suspend fun fileExists(path: String): Boolean = withContext(Dispatchers.IO) {
        val file = resolvePath(path)
        file.exists() && file.isFile
    }

    override suspend fun createDirectory(path: String) = withContext(Dispatchers.IO) {
        val dir = resolvePath(path)

        if (dir.exists()) {
            if (!dir.isDirectory) {
                throw SyncStorageException.IOError("$path exists but is not a directory")
            }
            // Already exists as directory - idempotent success
            return@withContext
        }

        try {
            if (!dir.mkdirs()) {
                throw SyncStorageException.IOError("Failed to create directory: $path")
            }
            Log.d(TAG, "Created directory: $path")
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        }
    }

    override suspend fun quarantineFile(path: String, reason: String): String = withContext(Dispatchers.IO) {
        val file = resolvePath(path)

        if (!file.exists()) {
            throw SyncStorageException.FileNotFound(path)
        }

        try {
            // Create quarantine directory
            val quarantineDir = resolvePath(SyncConstants.QUARANTINE_DIR)
            if (!quarantineDir.exists()) {
                quarantineDir.mkdirs()
            }

            // Generate quarantine filename with timestamp
            val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH-mm-ss", Locale.US)
            val timestamp = dateFormat.format(Date())
            val quarantineName = "${timestamp}_${file.name}"
            val quarantineFile = File(quarantineDir, quarantineName)

            // Move file to quarantine
            if (!file.renameTo(quarantineFile)) {
                file.copyTo(quarantineFile, overwrite = true)
                file.delete()
            }

            val quarantinePath = "${SyncConstants.QUARANTINE_DIR}/$quarantineName"
            Log.w(TAG, "Quarantined $path -> $quarantinePath (reason: $reason)")

            quarantinePath
        } catch (e: SecurityException) {
            throw SyncStorageException.PermissionDenied(path)
        } catch (e: IOException) {
            throw SyncStorageException.IOError("Failed to quarantine $path", e)
        }
    }

    override suspend fun watchForChanges(callback: (String) -> Unit): Closeable {
        // Note: Android doesn't have a built-in recursive file watcher like iOS/macOS.
        // For production, consider using FileObserver (limited) or polling.
        // This implementation provides a manual trigger mechanism.

        val isActive = AtomicBoolean(true)

        Log.d(TAG, "File watching registered for $rootPath")

        return object : Closeable {
            override fun close() {
                isActive.set(false)
                Log.d(TAG, "File watching stopped for $rootPath")
            }
        }
    }

    /**
     * Checks if the sync root folder is accessible.
     *
     * @return true if the folder exists and is readable/writable
     */
    suspend fun isAccessible(): Boolean = withContext(Dispatchers.IO) {
        try {
            rootFile.exists() && rootFile.isDirectory && rootFile.canRead() && rootFile.canWrite()
        } catch (e: SecurityException) {
            false
        }
    }

    /**
     * Gets the total size of all files in the sync folder.
     *
     * @return Total size in bytes
     */
    suspend fun getTotalSize(): Long = withContext(Dispatchers.IO) {
        try {
            rootFile.walkTopDown()
                .filter { it.isFile }
                .sumOf { it.length() }
        } catch (e: Exception) {
            Log.w(TAG, "Failed to calculate total size", e)
            0L
        }
    }

    /**
     * Gets the number of files in the sync folder.
     *
     * @return File count
     */
    suspend fun getFileCount(): Int = withContext(Dispatchers.IO) {
        try {
            rootFile.walkTopDown()
                .filter { it.isFile && !it.name.startsWith(".") }
                .count()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to count files", e)
            0
        }
    }
}

/**
 * Factory for creating LocalFolderSyncStorage instances.
 *
 * Validates the folder path and checks accessibility before creating storage.
 */
object LocalFolderSyncStorageFactory {
    /**
     * Creates a LocalFolderSyncStorage for the given path.
     *
     * @param path Absolute path to the sync folder
     * @return Storage instance
     * @throws IllegalArgumentException if path is invalid
     * @throws SyncStorageException.PermissionDenied if folder is not accessible
     */
    suspend fun create(path: String): LocalFolderSyncStorage = withContext(Dispatchers.IO) {
        val file = File(path)

        if (!file.exists()) {
            // Try to create the folder
            if (!file.mkdirs()) {
                throw SyncStorageException.IOError("Cannot create sync folder: $path")
            }
        }

        if (!file.isDirectory) {
            throw IllegalArgumentException("Path is not a directory: $path")
        }

        if (!file.canRead() || !file.canWrite()) {
            throw SyncStorageException.PermissionDenied(path)
        }

        LocalFolderSyncStorage(path)
    }

    /**
     * Creates a LocalFolderSyncStorage from a content URI.
     *
     * For use with Storage Access Framework (SAF) document trees.
     * Note: SAF URIs require special handling - this is a placeholder for
     * future implementation using DocumentFile API.
     *
     * @param uri Content URI from SAF folder picker
     * @return Storage instance
     */
    fun createFromUri(uri: Uri): LocalFolderSyncStorage {
        // For SAF URIs, we need to use DocumentFile API instead of java.io.File.
        // This is a placeholder - full SAF support would require a separate
        // DocumentFileSyncStorage implementation.
        throw UnsupportedOperationException(
            "SAF URI support not yet implemented. Use a direct file path instead."
        )
    }
}

File: Packages/BioMedLit/Sources/BioMedLit/Sync/LocalFolderSyncStorage.swift
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2025 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import Foundation

/// Sync storage backed by a local file system directory.
///
/// This implementation is useful for:
/// - Unit and integration testing
/// - LAN sync via shared network folders
/// - Debugging sync issues locally
/// - Development without cloud dependencies
///
/// Thread Safety: This class is an actor, providing automatic isolation
/// for all mutable state. All methods can be called safely from any context.
///
/// Example:
/// ```swift
/// let storage = try LocalFolderSyncStorage(
///     rootURL: URL(fileURLWithPath: "/path/to/sync")
/// )
/// try await storage.writeFile(data, at: "changes/device1/000001.json")
/// let files = try await storage.listFiles(at: "changes/device1")
/// ```
public actor LocalFolderSyncStorage: SyncStorageProtocol {
    // MARK: - Properties

    /// Root directory for sync files.
    private let rootURL: URL

    /// File manager instance (thread-safe for read operations).
    private let fileManager = FileManager.default

    /// ISO8601 date formatter for quarantine timestamps.
    private let dateFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    // MARK: - Initialization

    /// Creates local folder storage.
    ///
    /// Creates the root directory if it doesn't exist.
    ///
    /// - Parameter rootURL: Root directory for sync files.
    /// - Throws: `SyncStorageError.ioError` if directory cannot be created.
    public init(rootURL: URL) throws {
        self.rootURL = rootURL

        // Ensure root exists
        if !FileManager.default.fileExists(atPath: rootURL.path) {
            do {
                try FileManager.default.createDirectory(
                    at: rootURL,
                    withIntermediateDirectories: true
                )
            } catch {
                throw SyncStorageError.ioError(
                    "Failed to create sync root: \(error.localizedDescription)"
                )
            }
        }
    }

    // MARK: - Private Helpers

    /// Resolves a relative path to a full URL.
    ///
    /// - Parameter path: Relative path from sync root.
    /// - Returns: Absolute URL to the file or directory.
    private func resolve(_ path: String) -> URL {
        // Handle empty path
        if path.isEmpty {
            return rootURL
        }
        return rootURL.appendingPathComponent(path)
    }

    // MARK: - SyncStorageProtocol Implementation

    public func listFiles(at path: String) async throws -> [SyncFileInfo] {
        let url = resolve(path)

        // Check directory exists
        var isDirectory: ObjCBool = false
        guard fileManager.fileExists(atPath: url.path, isDirectory: &isDirectory),
              isDirectory.boolValue else {
            throw SyncStorageError.directoryNotFound(path)
        }

        // Get directory contents
        let contents: [URL]
        do {
            contents = try fileManager.contentsOfDirectory(
                at: url,
                includingPropertiesForKeys: [.fileSizeKey, .contentModificationDateKey],
                options: [.skipsHiddenFiles]
            )
        } catch {
            throw SyncStorageError.ioError(
                "Failed to list directory: \(error.localizedDescription)"
            )
        }

        // Filter to files only and extract metadata
        return try contents.compactMap { itemURL -> SyncFileInfo? in
            var isDir: ObjCBool = false
            fileManager.fileExists(atPath: itemURL.path, isDirectory: &isDir)

            // Skip directories
            guard !isDir.boolValue else { return nil }

            // Get file attributes
            let resourceValues: URLResourceValues
            do {
                resourceValues = try itemURL.resourceValues(
                    forKeys: [.fileSizeKey, .contentModificationDateKey]
                )
            } catch {
                throw SyncStorageError.ioError(
                    "Failed to read file attributes: \(error.localizedDescription)"
                )
            }

            return SyncFileInfo(
                name: itemURL.lastPathComponent,
                path: path.isEmpty ? itemURL.lastPathComponent : path + "/" + itemURL.lastPathComponent,
                size: resourceValues.fileSize ?? 0,
                modifiedAt: resourceValues.contentModificationDate ?? Date()
            )
        }
    }

    public func listDirectories(at path: String) async throws -> [String] {
        let url = resolve(path)

        // Return empty array if directory doesn't exist (not an error)
        guard fileManager.fileExists(atPath: url.path) else {
            return []
        }

        // Get directory contents
        let contents: [URL]
        do {
            contents = try fileManager.contentsOfDirectory(
                at: url,
                includingPropertiesForKeys: nil,
                options: [.skipsHiddenFiles]
            )
        } catch {
            throw SyncStorageError.ioError(
                "Failed to list directory: \(error.localizedDescription)"
            )
        }

        // Filter to directories only
        return contents.compactMap { itemURL -> String? in
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: itemURL.path, isDirectory: &isDirectory)
            return isDirectory.boolValue ? itemURL.lastPathComponent : nil
        }
    }

    public func readFile(at path: String) async throws -> Data {
        let url = resolve(path)

        guard fileManager.fileExists(atPath: url.path) else {
            throw SyncStorageError.fileNotFound(path)
        }

        do {
            return try Data(contentsOf: url)
        } catch {
            throw SyncStorageError.ioError(
                "Failed to read file: \(error.localizedDescription)"
            )
        }
    }

    public func writeFile(_ data: Data, at path: String) async throws {
        let url = resolve(path)

        // Ensure parent directory exists
        let parentURL = url.deletingLastPathComponent()
        if !fileManager.fileExists(atPath: parentURL.path) {
            do {
                try fileManager.createDirectory(
                    at: parentURL,
                    withIntermediateDirectories: true
                )
            } catch {
                throw SyncStorageError.ioError(
                    "Failed to create parent directory: \(error.localizedDescription)"
                )
            }
        }

        // Write atomically to prevent partial files
        do {
            try data.write(to: url, options: .atomic)
        } catch {
            throw SyncStorageError.ioError(
                "Failed to write file: \(error.localizedDescription)"
            )
        }
    }

    public func deleteFile(at path: String) async throws {
        let url = resolve(path)

        // Silently succeed if file doesn't exist (idempotent)
        guard fileManager.fileExists(atPath: url.path) else {
            return
        }

        do {
            try fileManager.removeItem(at: url)
        } catch {
            throw SyncStorageError.ioError(
                "Failed to delete file: \(error.localizedDescription)"
            )
        }
    }

    public func fileExists(at path: String) async -> Bool {
        // True for files and directories alike, matching the protocol
        // contract and iCloudSyncStorage — sync callers probe change-log
        // directories as well as individual files.
        fileManager.fileExists(atPath: resolve(path).path)
    }

    public func createDirectory(at path: String) async throws {
        let url = resolve(path)

        // Silently succeed if already exists (idempotent)
        if fileManager.fileExists(atPath: url.path) {
            return
        }

        do {
            try fileManager.createDirectory(
                at: url,
                withIntermediateDirectories: true
            )
        } catch {
            throw SyncStorageError.ioError(
                "Failed to create directory: \(error.localizedDescription)"
            )
        }
    }

    public func quarantineFile(at path: String, reason: String) async throws -> String {
        let sourceURL = resolve(path)

        // Check source exists
        guard fileManager.fileExists(atPath: sourceURL.path) else {
            throw SyncStorageError.fileNotFound(path)
        }

        // Generate quarantine path with timestamp
        let timestamp = dateFormatter.string(from: Date())
            .replacingOccurrences(of: ":", with: "-") // Windows-safe
        let quarantinePath = "\(SyncConstants.quarantineDirectory)/\(timestamp)_\(sourceURL.lastPathComponent)"
        let quarantineURL = resolve(quarantinePath)

        // Ensure quarantine directory exists
        try await createDirectory(at: SyncConstants.quarantineDirectory)

        // Move file to quarantine
        do {
            try fileManager.moveItem(at: sourceURL, to: quarantineURL)
        } catch {
            throw SyncStorageError.ioError(
                "Failed to quarantine file: \(error.localizedDescription)"
            )
        }

        // Write reason file for analysis
        let reasonPath = quarantinePath + ".reason.txt"
        let reasonContent = """
            Quarantine Reason: \(reason)
            Original Path: \(path)
            Quarantined At: \(Date())
            """
        if let reasonData = reasonContent.data(using: .utf8) {
            try await writeFile(reasonData, at: reasonPath)
        }

        return quarantinePath
    }

    public func watchForChanges(
        _ callback: @escaping @Sendable (String) -> Void
    ) async -> Any {
        // For local folders, use DispatchSource for file monitoring.
        // Note: This is a simplified implementation that monitors the root.
        // A production implementation would use FSEvents on macOS for
        // recursive monitoring.

        let descriptor = open(rootURL.path, O_EVTONLY)
        guard descriptor >= 0 else {
            // Return a no-op token if we can't open the directory
            return NSObject()
        }

        let source = DispatchSource.makeFileSystemObjectSource(
            fileDescriptor: descriptor,
            eventMask: [.write, .delete, .rename, .extend],
            queue: .global(qos: .utility)
        )

        let rootPath = rootURL.path
        source.setEventHandler {
            callback(rootPath)
        }

        source.setCancelHandler {
            close(descriptor)
        }

        source.resume()
        return source
    }

    // MARK: - Additional Convenience Methods

    /// Gets the absolute path to the sync root.
    ///
    /// Useful for debugging and logging.
    public var rootPath: String {
        rootURL.path
    }

    /// Checks if the storage root exists and is accessible.
    ///
    /// - Returns: True if the root directory exists and is readable.
    public func isAccessible() -> Bool {
        fileManager.isReadableFile(atPath: rootURL.path)
    }
}
